### PR TITLE
Make ModalContainer look for the parent lifecycle using ViewTreeLifecycleOwner.

### PR DESCRIPTION
**Pull request by zach-klippenstein**
This stops the leak of #314, but doesn't implement the full fix agreed upon there.

This change also changes the API a bit so it doesn't pollute the entire auto-complete
namespace for `Context` (or `View`) for consumers, which is important because I'm about
to add some more functions there and Workflow isn't an "AndroidX support support library".

### CODE REVIEW

File: workflow-ui/modal-android/src/main/java/com/squareup/workflow1/ui/modal/ModalContainer.kt [hunk 1]
```diff
@@ -76,7 +76,9 @@ public abstract class ModalContainer<ModalRenderingT : Any> @JvmOverloads constr
                 override fun onViewAttachedToWindow(v: View) {
                   // Note this is a different lifecycle than the WorkflowLifecycleOwner – it will
                   // probably be the owning AppCompatActivity.
-                  lifecycle = ref.dialog.decorView?.context?.lifecycleOrNull()
+                  lifecycle = ref.dialog.decorView
+                    ?.let(::lifecycleOwnerFromViewTreeOrContext)
```
Comment (zach-klippenstein) [inline workflow-ui/modal-android/src/main/java/com/squareup/workflow1/ui/modal/ModalContainer.kt:80]: TODO unit tests that this actually fixes #314

**Review by rjrjr — APPROVED** [review]

**Comment by zach-klippenstein** [comment]
@rjrjr I know you already approved, but checkout the tests this broke: https://github.com/square/workflow-kotlin/compare/43ed69687a44022eb09a9826cccbb99dae2f5cc8..cdca31cecf62accc50c1594f45e3fa622d47a328

We could have been leaking a lot.

**Comment by rjrjr** [comment]
Yikes.

**Comment by zach-klippenstein** [comment]
The failure on API 21 looks like something that could be legit:
```
com.squareup.workflow1.ui.backstack.test.BackstackContainerTest > lifecycle_replace_after_pause[test(AVD) - 5.0.2] FAILED 
	java.lang.RuntimeException: missing (1)                 : LeafView next ON_START
	---
```

However I can't repro it locally on 21 or 29.

The failure on 29 looks like a flake.

**Pull request merged**